### PR TITLE
Add setup for config, logging, and plugins

### PR DESCRIPTION
**Pull request by econchick**
**What this PR does / why we need it**:

This is copy-pasta from gordon core. Rather than create a common package immediately, I chose to hold off to see if the plugin logic will need to be different in some way. Once the janitor is further develop, we can re-evaluate the need for a gordon-common package.

**Which issue(s) this PR fixes**

N/A

**Special notes for your reviewer**:

**Release note**:

```release-note
None
```

cc @spotify/alf

### CODE REVIEW

File: gordon_janitor/main.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        A dict for Gordon service configuration
+    """
+    config = _load_config(root=config_root)
+
+    logging_config = config.get('core', {}).get('logging', {})
+    log_level = logging_config.get('level', 'INFO').upper()
+    log_handlers = logging_config.get('handlers') or ['syslog']
+
+    ulogger.setup_logging(
+        progname='gordon', level=log_level, handlers=log_handlers)
```
Comment (matt0bi) [inline gordon_janitor/main.py]: What do you think about changing `progname='gordon'` to `progname='gordon-janitor'` to make the logs more explicit?

File: gordon_janitor/plugins_loader.py [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+        # snip
+        entry_points={
+            'gordon_janitor.plugins': [
+                'gcp.gdns = gordon_janitor_gcp.gdns:DNSProviderClient',
+            ],
+        },
+        # snip
+    )
+
+Plugins are initialized with any config defined in ``gordon-user.toml``
```
Comment (matt0bi) [inline gordon_janitor/plugins_loader.py]: `gordon-user.toml` and `gordon.toml` should be `gordon-janitor-user.toml` and `gordon-janitor.toml` to be in line with `main.py`.

File: gordon_janitor/plugins_loader.py [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+and ``gordon.toml``. See :doc:`config` for more details.
+
+Once a plugin is found, the loader looks up its configuration via the
+same key defined in its entry point, e.g. ``gcp.gpubsub``.
+
+The value of the entry point (e.g. ``gordon_gcp.gpubsub:EventClient``)
+must point to a class. The plugin class is instantiated with its config.
+
+A plugin will not have access to another plugin's configuration. For
+example, the ``gcp.gpusub`` will not have access to the configuration
```
Comment (matt0bi) [inline gordon_janitor/plugins_loader.py]: Nitpick, but '_the `gcp.gpubsub` will_' should either drop the '_the_' or specify something after the plugin name eg. '_the `gcp.gpubsub` plugin_'. There's also a missing letter in `gcp.gpusub` (if it was supposed to be `gcp.gpubsub`, it not, never mind).

File: gordon_janitor/plugins_loader.py [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+for ``gcp.gdns``.
+
+See :doc:`plugins` for details on how to write a plugin for Gordon.
```
Comment (matt0bi) [inline gordon_janitor/plugins_loader.py]: Another nitpick, but should '_plugin for Gordon._' be '_plugin for Gordon-Janitor._'?

File: MANIFEST.in [hunk 5]
```diff
@@ -3,7 +3,7 @@ recursive-include gordon_janitor *.py
 recursive-include tests *.py
 
 # Exclude service-specific config
-exclude gordon*.toml
+exclude gordon-janitor*.toml
```
Comment (matt0bi) [inline MANIFEST.in:6]: Good idea :+1:

File: gordon_janitor/plugins_loader.py [hunk 6]
```diff
@@ -0,0 +1,206 @@
...
+        gathered_plugins[entry_point.name] = entry_point
+    return gathered_plugins
+
+
+def load_plugins(config):
+    """
+    Discover and instantiate plugins.
+
+    Args:
+        config (dict): loaded configuration for the Gordon service.
```
Comment (matt0bi) [inline gordon_janitor/plugins_loader.py]: Another word nitpick: `s/Gordon/Gordon-Janitor/` here and in `test_main.py` on line 119 and possibly on line 197 in the Returns part of this docstring ('_activated in gordon config._' -> '_activated in gordon-janitor config._`.

**Review by matt0bi — CHANGES_REQUESTED** [review]
Overall looks good, I had a few naming related nitpicks (eg. gordon -> gordon-janitor).

**Comment by econchick** [comment]
@matt0bi updated - mind taking a quick look?

**Review by matt0bi — APPROVED** [review]

**Comment by matt0bi** [comment]
🚢

**Pull request merged**